### PR TITLE
add sign-in component

### PR DESCRIPTION
**Pull request by 928PJY**

### CODE REVIEW

File: test/unitTests/credential/CredentialController.test.ts [hunk 1]
```diff
@@ -0,0 +1,347 @@
+import * as vscode from 'vscode';
+import { CredentialExpiry, ResetCredential, RefreshCredential, EnvironmentChange, BaseEvent, FetchFromLocalCredentialManager, UserSignedOut, LogProgress, UserSigningIn, UserSignedIn, SignInFailed } from '../../../src/common/loggingEvents';
+import { EventStream } from '../../../src/common/EventStream';
+import { CredentialController, Credential } from '../../../src/credential/CredentialController';
+import { KeyChain } from '../../../src/credential/KeyChain';
+import { EnvironmentController } from '../../../src/common/EnvironmentController';
+// import { UserInfo } from '../../../src/common/shared';
```
Comment (928PJY) [inline test/unitTests/credential/CredentialController.test.ts]: remove #Resolved

File: src/common/shared.ts [hunk 2]
```diff
@@ -12,5 +12,27 @@ export const PACKAGE_JSON = extension.packageJSON;
...
+export type Environment = 'PROD' | 'PPE';
+
+export class MessageAction implements vscode.MessageItem {
+    constructor(public title: string, public command?: string, public desctiption?: string, public callback?: Function, public args?: any[]) { }
+}
+
+// OAuth
+export type DocsSignInStatus = 'Initializing' | 'SigningIn' | 'SignedIn' | 'SignedOut';
+
+export type DocsSignInType = 'GitHub' | 'DevOps';
```
Comment (herohua) [inline src/common/shared.ts]: `AzureDevOps`? #Resolved

File: src/credential/CredentialController.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    public async signIn(): Promise<void> {
+        try {
+            // Step-0: Start login
+            this.resetCredential();
+            this.signInStatus = 'SigningIn';
+            this.eventStream.post(new UserSigningIn());
+
+            // Step-1: AAD
+            if (!this.aadInfo) {
+                this.eventStream.post(new LogProgress(`Sign in to docs build with AAD...`, 'Sign In'));
```
Comment (herohua) [inline src/credential/CredentialController.ts]: `Sign-in to Docs build with AAD...` #Resolved

File: src/credential/CredentialController.ts [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+                if (result instanceof Error) {
+                    throw result;
+                }
+                return result;
+            }
+            this.eventStream.post(new SignInFailed(`Sign In with AAD Failed`));
+            return undefined;
+        } catch (err) {
+            this.eventStream.post(new SignInFailed(`Sign In with AAD Failed: ${err.message}`));
```
Comment (herohua) [inline src/credential/CredentialController.ts]: `Sign-in` #Resolved

File: src/credential/CredentialController.ts [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+                        reject(err);
+                    }
+                });
+
+                if (result instanceof Error) {
+                    throw result;
+                }
+                return result;
+            }
+            this.eventStream.post(new SignInFailed(`Sign In with AAD Failed`));
```
Comment (herohua) [inline src/credential/CredentialController.ts]: `Sign-in` #Resolved

File: src/observers/DocsLoggerObserver.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    private handleUserSignedIn(event: UserSignedIn) {
+        this.appendLine(`Successfully sign in to Docs build system:`);
+        this.appendLine(`    - GitHub Acount: ${event.credential.userInfo.userName}`);
+        this.appendLine(`    - User email   : ${event.credential.userInfo.userEmail}`);
+        this.appendLine();
+    }
+
+    private handleFetchFromLocalCredentialManager(event: FetchFromLocalCredentialManager) {
+        this.appendLine(`Successfully get User credential From Local Credential Manager:`);
+        this.appendLine(`    - GitHub Acount: ${event.credential.userInfo.userName}`);
```
Comment (herohua) [inline src/observers/DocsLoggerObserver.ts]: `GitHub account` #Resolved

File: src/observers/DocsLoggerObserver.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    // Sign
+    private handleUserSignedIn(event: UserSignedIn) {
+        this.appendLine(`Successfully sign in to Docs build system:`);
+        this.appendLine(`    - GitHub Acount: ${event.credential.userInfo.userName}`);
+        this.appendLine(`    - User email   : ${event.credential.userInfo.userEmail}`);
+        this.appendLine();
+    }
+
+    private handleFetchFromLocalCredentialManager(event: FetchFromLocalCredentialManager) {
+        this.appendLine(`Successfully get User credential From Local Credential Manager:`);
```
Comment (herohua) [inline src/observers/DocsLoggerObserver.ts]: `Successfully retrieved user credential from Local Credential Manager:` #Resolved

File: src/observers/ErrorMessageObserver.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+import { EventType } from "../common/EventType";
+import { BaseEvent, SignInFailed } from '../common/loggingEvents';
+import { MessageAction } from '../common/shared';
+
+export class ErrorMessageObserver {
+    public eventHandler = (event: BaseEvent) => {
+        switch (event.type) {
+            case EventType.SignInFailed:
+                let asSigninFailed = <SignInFailed>event;
+                this.showErrorMessage(`SignIn failed: ${asSigninFailed.message}`);
```
Comment (herohua) [inline src/observers/ErrorMessageObserver.ts]: `Sign-in` #Resolved

File: src/observers/ErrorMessageObserver.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+            if (input.command) {
+                vscode.commands.executeCommand(input.command, undefined);
+            } else if (input.callback) {
+                input.callback(input.args);
+            }
+        }
+    }
+
+    private handleCredentialExpiry() {
+        let message = `The Credential has expiried, please Re-Sign to refresh the credentail`;
```
Comment (herohua) [inline src/observers/ErrorMessageObserver.ts]: `Credential has expired. Please sign-in again to continue.` #Resolved

File: src/observers/InfoMessageObserver.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import * as vscode from 'vscode';
+import { BaseEvent } from "../common/loggingEvents";
+import { EventType } from "../common/EventType";
+import { MessageAction } from '../common/shared';
+export class InfoMessageObserver {
+    public eventHandler = (event: BaseEvent) => {
+        switch (event.type) {
+            case EventType.UserSignedIn:
+                this.showInfoMessage('Sign In Successfully!', new MessageAction('Build', 'docs.build', 'Do you want to build Current workspace folder?'));
```
Comment (herohua) [inline src/observers/InfoMessageObserver.ts]: `Sign-in successfully!` #Resolved

**Comment by 928PJY** [comment]
Split to small PR: #9  and #10 
Abadon this PR

**Pull request closed without merging**